arangod/Graph/Providers/ClusterProvider.cpp

```diff
@@ -0,0 +1,295 @@
+  std::vector<Step*> result{};
+
+  if (looseEnds.size() > 0) {
+    result.reserve(looseEnds.size());
+    auto const* engines = _opts.getCache()->engines();
+
+    // slow path, sharding not deducable from _id
+    transaction::BuilderLeaser leased(trx());
+    leased->openObject();
+    leased->add("keys", VPackValue(VPackValueType::Array));
+    for (auto const& looseEnd : looseEnds) {
+      LOG_DEVEL << " -> " << looseEnd->toString();
+      leased->add(VPackValuePair(looseEnd->getVertex().getID().data(),
+                                 looseEnd->getVertex().getID().length(),
+                                 VPackValueType::String));
+    }
+    leased->close();  // 'keys' Array
+    leased->close();  // base object
+
+    auto* pool = trx()->vocbase().server().getFeature<NetworkFeature>().pool();
+
+    network::RequestOptions reqOpts;
+    reqOpts.database = trx()->vocbase().name();
+    reqOpts.skipScheduler = true;  // hack to avoid scheduler queue
+
+    std::vector<Future<network::Response>> futures;
+    futures.reserve(engines->size());
+
+    for (auto const& engine : *engines) {
+      futures.emplace_back(
+          network::sendRequestRetry(pool, "server:" + engine.first, fuerte::RestVerb::Put,
+                                    ::vertexUrl + StringUtils::itoa(engine.second),
+                                    leased->bufferRef(), reqOpts));
+    }
+
+    for (Future<network::Response>& f : futures) {
+      network::Response const& r = f.get();
+
+      if (r.fail()) {
+        THROW_ARANGO_EXCEPTION(network::fuerteToArangoErrorCode(r));
+      }
+
+      auto payload = r.response().stealPayload();
+      VPackSlice resSlice(payload->data());
+      if (!resSlice.isObject()) {
+        // Response has invalid format
+        THROW_ARANGO_EXCEPTION(TRI_ERROR_HTTP_CORRUPTED_JSON);
+      }
+      if (r.statusCode() != fuerte::StatusOK) {
+        // We have an error case here. Throw it.
+        THROW_ARANGO_EXCEPTION(network::resultFromBody(resSlice, TRI_ERROR_INTERNAL));
+      }
+
+      bool cached = false;
+      for (auto pair : VPackObjectIterator(resSlice, /*sequential*/ true)) {
+        arangodb::velocypack::HashedStringRef key(pair.key);
+        /*if (ADB_UNLIKELY(vertexIds.erase(key) == 0)) {
+         // TODO: Check!
+          // We either found the same vertex twice,
+          // or found a vertex we did not request.
+          // Anyways something somewhere went seriously wrong
+          THROW_ARANGO_EXCEPTION(TRI_ERROR_CLUSTER_GOT_CONTRADICTING_ANSWERS);
+        }*/
+
+        // TODO: Check!
+        // TRI_ASSERT(result.find(key) == result.end());
+        if (!cached) {
+          _opts.getCache()->datalake().add(std::move(payload));
+          cached = true;
+        }
+
+        // Protected by datalake
+        LOG_DEVEL << "Found key: " << key.toString();
+        LOG_DEVEL << pair.value.toString();
+
+        // TODO: This needs to be optimised!!
```

Comment:
Logically correct, but needs to be optimized. Discuss how to access/find a Step fast.